compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTensorPad.cpp

```diff
@@ -0,0 +1,178 @@
+// Copyright 2022 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Codegen/Common/LinalgOpInfo.h"
+#include "iree/compiler/Codegen/LLVMGPU/TransposeUtils.h"
+#include "iree/compiler/Codegen/PassDetail.h"
+#include "iree/compiler/Codegen/Passes.h"
+#include "iree/compiler/Dialect/Flow/IR/FlowOps.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Bufferization/IR/Bufferization.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#define DEBUG_TYPE "iree-llvmgpu-tensor-pad"
+
+namespace mlir {
+namespace iree_compiler {
+
+namespace {
+
+static FailureOr<SmallVector<Value>> rewriteAsPaddedOp(
+    PatternRewriter &rewriter, linalg::LinalgOp linalgOp,
+    linalg::LinalgOp &paddedOp) {
+  Location loc = linalgOp.getLoc();
+
+  PatternRewriter::InsertionGuard g(rewriter);
+  // Set IP after op because we also take the dims of the original output.
+  rewriter.setInsertionPointAfter(linalgOp);
+
+  // Pad each input operand in shared memory out to 32,32
+  SmallVector<Value> paddedOperands;
+  paddedOperands.reserve(linalgOp.getNumInputsAndOutputs());
+  for (OpOperand *opOperand : linalgOp.getInputAndOutputOperands()) {
+    // Find DispatchTensorLoadOp's feeding into the linalg or abort.
+    auto tensorLoad = dyn_cast_or_null<IREE::Flow::DispatchTensorLoadOp>(
+        opOperand->get().getDefiningOp());
+    if (!tensorLoad) {
+      return rewriter.notifyMatchFailure(linalgOp, "does not have tensor load");
+    }
+
+    // Determine the padded shape from the load
+    ArrayRef<int64_t> shape = linalgOp.getShape(opOperand);
+    SmallVector<int64_t> paddedShape(shape.begin(), shape.end());
+    for (const auto &en : llvm::enumerate(tensorLoad.getMixedSizes())) {
+      if (en.value().is<Attribute>()) {
+        paddedShape[en.index()] =
+            en.value().get<Attribute>().dyn_cast<IntegerAttr>().getInt();
+      } else {
```

Comment:
The check of dyn_cast<IntegerAttr> is missing, maybe use `getConstantIntValue` method, then we don't need to check ourself?

```mlir
Optional<int64_t> cst = getConstantIntValue(en.value());
...
```